pkg/pillar/zedcloud/authen.go

```diff
@@ -0,0 +1,508 @@
+			return nil, types.ExtStatusSignVerifyFail, err
+		}
+		log.Infof("verifyAuthentication: ok\n")
+	}
+	return data, extraSt, nil
+}
+
+func getCloudCert() (*x509.Certificate, error) {
+	var sCert *x509.Certificate
+	if serverLeafCert == nil {
+		certBytes, err := ioutil.ReadFile(types.ServerCertName)
+		if err != nil {
+			log.Infof("getCloudCert: can not read in server cert file, %v\n", err)
+			return sCert, err
+		}
+		block, _ := pem.Decode(certBytes)
+		if block == nil {
+			err := fmt.Errorf("getCloudCert: can not get client Cert")
+			return sCert, err
+		}
+
+		sCert, err = x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			log.Infof("getCloudCert: can not parse cert %v\n", err)
+			return sCert, err
+		}
+
+		// hash verify using PEM bytes from cloud
+		cloudCertHash = ComputeSha(certBytes)
+		serverLeafCert = sCert
+
+	} else {
+		sCert = serverLeafCert
+	}
+
+	return sCert, nil
+}
+
+// verify the signed data with cloud certificate public key
+func verifyAuthSig(signature []byte, cert *x509.Certificate, hash []byte) error {
+
+	log.Debugf("verifyAuthsig sigdata (len %d) %v\n", len(hash), hash)
+
+	switch pub := cert.PublicKey.(type) {
+	case *rsa.PublicKey:
+		err := rsa.VerifyPKCS1v15(pub, crypto.SHA256, hash, signature)
+		if err != nil {
+			return err
+		}
+		log.Infof("verifyAuthSig: verify rsa ok\n")
+	case *ecdsa.PublicKey:
+
+		sigHalflen := len(signature) / 2
+		rbytes := signature[0:sigHalflen]
+		sbytes := signature[sigHalflen:]
+		r := new(big.Int)
+		s := new(big.Int)
+		r.SetBytes(rbytes)
+		s.SetBytes(sbytes)
+		ok := ecdsa.Verify(pub, hash, r, s)
+		if !ok {
+			return errors.New("ecdsa image signature verification failed")
+		}
+		log.Infof("verifyAuthSig: verify ecdsa ok\n")
+	default:
+		log.Debugf("verifyAuthSig: unknown type of public key\n")
+		return errors.New("unknown type of public key")
+	}
+	return nil
+}
+
+// add an authentication envelope protobuf when sending http POST
+func addAuthentication(ctx ZedCloudContext, b *bytes.Buffer, useOnboard bool) (*bytes.Buffer, error) {
+	var data []byte
+	if b != nil {
+		data = b.Bytes()
+	}
+	body := zauth.AuthBody{
+		Payload: data,
+	}
+	sm := zauth.AuthContainer{}
+	sm.AuthPayload = &body
+
+	cert, err := getMyDevCert(useOnboard)
+	if err != nil {
+		log.Debugf("addAuthenticate: get client cert failed\n")
+		return nil, err
+	}
+
+	// assign our certificate hash of 32 bytes
+	if useOnboard {
+		sm.SenderCertHash = onBoardCertHash
+		sm.SenderCert = onBoardCertBytes
+		if len(sm.SenderCert) == 0 {
+			err := fmt.Errorf("addAuthentication: SenderCert empty")
+			log.Infof("addAuthenticate: get sender cert failed, %v\n", err)
+			return nil, err
+		}
+		log.Infof("addAuthenticate: onboard senderCert size %d\n", len(sm.SenderCert))
+	} else {
+		sm.SenderCertHash = deviceCertHash
+	}
+	if sm.SenderCertHash == nil {
+		err := fmt.Errorf("addAuthentication: SenderCertHash empty")
+		return nil, err
+	}
+
+	sig, err := signAuthData(data, cert)
```

Comment:
This returns an unspecified number of bytes. Should we explicitly truncate (to 32 currently) to make it more obvious how different lengths would be supported?

Reply:
this length i found is 64: "signData: ecdas sigres (len 64): 7b09cd14792ca7687548de83667081c97f1f98e9c89c2526b0b21359c038b3cb83019" from printout

Reply:
Oh my confusion between the hash and the signature.